### PR TITLE
(Refactor) Arithmetic monomial sum

### PR DESCRIPTION
**Pull request by ajreynol**
This copies the monomial sum utility from quantifiers/ and puts it in arith/, changing its name.

Most references to QuantArith have changed to ArithMSum. However, to avoid merge conflicts, some of the references to the old copy (QuantArith) were kept.  We need to merge this PR first, for instance, to satisfy the review on #1221.  

Thus, after this PR is merged, I will update the references to QuantArith -> ArithMSum when master is merged into the respective PRs. When this is done, QuantArith will be deleted.

### CODE REVIEW

File: src/theory/arith/arith_msum.cpp [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+  else
+  {
+    return false;
+  }
+}
+bool ArithMSum::getMonomial(Node n, std::map<Node, Node>& msum)
+{
+  if (n.isConst())
+  {
+    if (msum.find(Node::null()) == msum.end())
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Is it possible to merge these two if conditions?

File: src/theory/arith/arith_msum.cpp [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    c = n[0];
+    v = n[1];
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+bool ArithMSum::getMonomial(Node n, std::map<Node, Node>& msum)
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Just to check, does `msum` need to be ordered?

File: src/theory/arith/arith_msum.cpp [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    }
+  }
+  return false;
+}
+
+bool ArithMSum::getMonomialSum(Node n, std::map<Node, Node>& msum)
+{
+  if (n.getKind() == PLUS)
+  {
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Use for each loop.

File: src/theory/arith/arith_msum.cpp [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+                  it->second.isNull()
+                      ? NodeManager::currentNM()->mkConst(Rational(1))
+                      : it->second);
+              msum[it->first] = Rewriter::rewrite(r);
+            }
+            else
+            {
+              msum[it->first] =
+                  it->second.isNull()
+                      ? NodeManager::currentNM()->mkConst(Rational(-1))
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Avoid redundant calls such as calling NodeManager::currentNM() in a for loop.

File: src/theory/arith/arith_msum.cpp [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+                    : NodeManager::currentNM()->mkConst(Rational(0)));
+}
+
+Node ArithMSum::mkCoeffTerm(Node coeff, Node t)
+{
+  if (coeff.isNull())
+  {
+    return t;
+  }
+  else
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Get rid of the else branch here.
Even more elegant would be:
```
return coeff.IsNull() ? t : NodeManager::currentNM()->mkNode(kind::MULT, coeff, t);
```

File: src/theory/arith/arith_msum.cpp [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+          if (!it->first.isNull())
+          {
+            if (!it->second.isNull())
+            {
+              m = NodeManager::currentNM()->mkNode(MULT, it->second, it->first);
+            }
+            else
+            {
+              m = it->first;
+            }
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: This if statement might also be easier to read if you use `m = ... ? ... : ....;`

File: src/theory/arith/arith_msum.cpp [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+           ++it)
+      {
+        if (it->first != v)
+        {
+          Node m;
+          if (!it->first.isNull())
+          {
+            if (!it->second.isNull())
+            {
+              m = NodeManager::currentNM()->mkNode(MULT, it->second, it->first);
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Again, calling currentNM() in a loop.

File: src/theory/arith/arith_msum.cpp [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        }
+      }
+      if (r.sgn() == 1)
+      {
+        val = negate(val);
+      }
+      else
+      {
+        val = Rewriter::rewrite(val);
+      }
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Again, imo easier to read if you use `val = ... ? ... : ... ;`

File: src/theory/arith/arith_msum.cpp [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+  }
+  if (tn.isReal())
+  {
+    std::map<Node, Node> msum;
+    if (ArithMSum::getMonomialSumLit(lit, msum))
+    {
+      Node val, veqc;
+      if (ArithMSum::isolate(v, msum, veqc, val, EQUAL) != 0)
+      {
+        if (veqc.isNull())
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Merge these two if conditions?

File: src/theory/arith/arith_msum.cpp [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    }
+    else
+    {
+      coeff = it->second;
+      msum.erase(v);
+      rem = mkNode(msum);
+      return true;
+    }
+  }
+  else
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Get rid of else branch.

File: src/theory/arith/arith_msum.h [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+   *
+   * If this function returns a value ret where ret != 0, then veq
+   * is set to a literal that is equivalent to ([msum] <k> 0), and:
+   *    If ret=1, then veq is of the form ( v <k> val) if veq_c.isNull(),
+   *                   or ([veq_c] * v <k> val) if !veq_c.isNull().
+   *   If ret=-1, then veq is of the form ( val <k> v) if veq_c.isNull(),
+   *                   or (val <k> [veq_c] * v) if !veq_c.isNull().
+   * If doCoeff = false or v does not have Integer type, then veq_c is null.
+   *
+   * This function returns 0 indiciating a failure if msum does not contain
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.h]: indicating (-i)

File: src/theory/quantifiers/ce_guided_single_inv_sol.cpp [hunk 12]
```diff
@@ -300,7 +300,7 @@ bool CegConjectureSingleInvSol::getAssignEquality( Node eq, std::vector< Node >&
   TypeNode tn = eq[0].getType();
   if( tn.isInteger() || tn.isReal() ){
     std::map< Node, Node > msum;
-    if( QuantArith::getMonomialSumLit( eq, msum ) ){
+    if( ArithMSum::getMonomialSumLit( eq, msum ) ){
```
Comment (aniemetz) [inline src/theory/quantifiers/ce_guided_single_inv_sol.cpp]: Format?

File: src/theory/arith/arith_msum.cpp [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+
+bool ArithMSum::getMonomial(Node n, Node& c, Node& v)
+{
+  if (n.getKind() == MULT && n.getNumChildren() == 2 && n[0].isConst())
+  {
+    c = n[0];
+    v = n[1];
+    return true;
+  }
+  else
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Get rid of else branch.

File: src/theory/arith/arith_msum.cpp [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
+    {
+      if (!getMonomial(n[i], msum))
+      {
+        return false;
+      }
+    }
+    return true;
+  }
+  else
```
Comment (aniemetz) [inline src/theory/arith/arith_msum.cpp]: Get rid of else branch.

File: src/theory/quantifiers/ce_guided_single_inv_sol.cpp (hunk 12 above)
Comment (ajreynol) [reply src/theory/quantifiers/ce_guided_single_inv_sol.cpp]: It's dead code, perhaps I should remove.

File: src/theory/arith/arith_msum.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/arith/arith_msum.cpp]: Yes, see line 49 of arith_msum.h

File: src/theory/arith/arith_msum.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/arith/arith_msum.cpp]: No, the top level if-then-else is splitting on the cases for n and the inner ifs are checking whether we can constructor an msum from n.

File: src/theory/arith/arith_msum.cpp (hunk 8 above)
Comment (ajreynol) [reply src/theory/arith/arith_msum.cpp]: I've made mkCoeffTerm inline and made these portions call that function.

File: src/theory/arith/arith_msum.cpp (hunk 9 above)
Comment (ajreynol) [reply src/theory/arith/arith_msum.cpp]: I don't like merging conditions where the order becomes critical (veqc is computed by the first condition)

File: src/theory/arith/arith_msum.h (hunk 11 above)
Comment (ajreynol) [reply src/theory/arith/arith_msum.h]: What do you mean here?

Comment (aniemetz) [reply src/theory/arith/arith_msum.h]: Your wrote indiciating instead of indicating.

Comment (ajreynol) [reply src/theory/arith/arith_msum.h]: Ah I was interpreting this as negation of quantity i (too much thinking about arithmetic)

**Pull request merged**